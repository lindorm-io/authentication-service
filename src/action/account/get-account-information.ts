import Joi from "@hapi/joi";
import { IAuthContext } from "../../typing";
import {
  IGetAccountDevicesData,
  IGetAccountSessionsData,
  getAccount,
  getAccountDevices,
  getAccountSessions,
} from "../../support";

export interface IGetAccountInformationOptions {
  accountId: string;
}

export interface IGetAccountInformationData {
  created: Date;
  devices: Array<IGetAccountDevicesData>;
  email: string;
  hasOtp: boolean;
  hasPassword: boolean;
  permission: string;
  sessions: Array<IGetAccountSessionsData>;
  updated: Date;
}

const schema = Joi.object({
  accountId: Joi.string().guid().required(),
});

export const getAccountInformation = (ctx: IAuthContext) => async (
  options: IGetAccountInformationOptions,
): Promise<IGetAccountInformationData> => {
  await schema.validateAsync(options);

  const { logger } = ctx;
  const { accountId } = options;

  logger.info("requesting account information", { id: accountId });

  const account = await getAccount(ctx)(accountId);
  const devices = await getAccountDevices(ctx)(account);
  const sessions = await getAccountSessions(ctx)(account);

  return {
    created: account.created,
    devices,
    email: account.email,
    hasOtp: !!account.otp.uri,
    hasPassword: !!account.password.signature,
    permission: account.permission,
    sessions,
    updated: account.updated,
  };
};
