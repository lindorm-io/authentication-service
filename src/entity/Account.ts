import { AccountEvent } from "../enum";
import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { Permission } from "@lindorm-io/jwt";

export interface IAccountOTP {
  signature: string;
  uri: string;
}

export interface IAccountPassword {
  signature: string;
  updated: Date;
}

export interface IAccount extends IEntity {
  email: string;
  identityLinked: boolean;
  otp: IAccountOTP;
  password: IAccountPassword;
  permission: string;
}

export interface IAccountOptions extends IEntityBaseOptions {
  email: string;
  identityLinked?: boolean;
  otp?: IAccountOTP;
  password?: IAccountPassword;
  permission?: string;
}

export class Account extends EntityBase implements IAccount {
  private _email: string;
  private _identityLinked: boolean;
  private _otp: IAccountOTP;
  private _password: IAccountPassword;
  private _permission: string;

  constructor(options: IAccountOptions) {
    super(options);
    this._email = options.email.toLowerCase();
    this._identityLinked = options.identityLinked || false;
    this._otp = {
      signature: options.otp?.signature || null,
      uri: options.otp?.uri || null,
    };
    this._password = {
      signature: options.password?.signature || null,
      updated: options.password?.updated || null,
    };
    this._permission = options.permission || Permission.USER;
  }

  public get email(): string {
    return this._email;
  }
  public set email(email: string) {
    this._email = email.toLowerCase();
    this.addEvent(AccountEvent.EMAIL_CHANGED, { email: this._email });
  }

  public get identityLinked(): boolean {
    return this._identityLinked;
  }
  public set identityLinked(identityLinked: boolean) {
    this._identityLinked = identityLinked;
    this.addEvent(AccountEvent.EMAIL_CHANGED, { identityLinked: this._identityLinked });
  }

  public get otp(): IAccountOTP {
    return this._otp;
  }
  public set otp(otp: IAccountOTP) {
    this._otp = otp;
    this.addEvent(AccountEvent.OTP_CHANGED, { otp: this._otp });
  }

  public get password(): IAccountPassword {
    return this._password;
  }
  public set password(password: IAccountPassword) {
    this._password = password;
    this.addEvent(AccountEvent.PASSWORD_CHANGED, { password: this._password });
  }

  public get permission(): string {
    return this._permission;
  }
  public set permission(permission: string) {
    this._permission = permission;
    this.addEvent(AccountEvent.PERMISSION_CHANGED, { permission: this._permission });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== AccountEvent.CREATED) continue;
      throw new EntityCreationError("Account");
    }
    this.addEvent(AccountEvent.CREATED, {
      email: this._email,
      permission: this._permission,
      password: this._password,
      otp: this._otp,
      created: this._created,
      updated: this._updated,
    });
  }
}
