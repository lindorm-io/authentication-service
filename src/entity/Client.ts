import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { ClientEvent } from "../enum";

export interface IClient extends IEntity {
  approved: boolean;
  description: string;
  emailAuthorizationUri: string;
  name: string;
  secret: string;
}

export interface IClientOptions extends IEntityBaseOptions {
  approved?: boolean;
  description?: string;
  emailAuthorizationUri?: string;
  name?: string;
  secret?: string;
}

export class Client extends EntityBase implements IClient {
  private _approved: boolean;
  private _description: string;
  private _emailAuthorizationUri: string;
  private _name: string;
  private _secret: string;

  constructor(options?: IClientOptions) {
    super(options);
    this._approved = options?.approved || false;
    this._description = options?.description || null;
    this._emailAuthorizationUri = options?.emailAuthorizationUri || null;
    this._name = options?.name || null;
    this._secret = options?.secret || null;
  }

  public get approved(): boolean {
    return this._approved;
  }
  public set approved(approved: boolean) {
    this._approved = approved;
    this.addEvent(ClientEvent.APPROVED_CHANGED, { approved: this._approved });
  }

  public get emailAuthorizationUri(): string {
    return this._emailAuthorizationUri;
  }
  public set emailAuthorizationUri(uri: string) {
    this._emailAuthorizationUri = uri;
    this.addEvent(ClientEvent.EMAIL_AUTHORIZATION_URI_CHANGED, { uri: this._emailAuthorizationUri });
  }

  public get description(): string {
    return this._description;
  }
  public set description(description: string) {
    this._description = description;
    this.addEvent(ClientEvent.DESCRIPTION_CHANGED, { description: this._description });
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(ClientEvent.NAME_CHANGED, { name: this._name });
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(ClientEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== ClientEvent.CREATED) continue;
      throw new Error("Client has already been created");
    }
    this.addEvent(ClientEvent.CREATED, {
      name: this._name,
      emailAuthorizationUri: this._emailAuthorizationUri,
      approved: this._approved,
      description: this._description,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
