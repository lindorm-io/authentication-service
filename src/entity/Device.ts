import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { DeviceEvent } from "../enum";

export interface IDevicePIN {
  signature: string;
  updated: Date;
}

export interface IDevice extends IEntity {
  accountId: string;
  name: string;
  pin: IDevicePIN;
  publicKey: string;
  secret: string;
}

export interface IDeviceOptions extends IEntityBaseOptions {
  accountId: string;
  name?: string;
  pin?: IDevicePIN;
  publicKey: string;
  secret?: string;
}

export class Device extends EntityBase implements IDevice {
  private _accountId: string;
  private _name: string;
  private _pin: IDevicePIN;
  private _publicKey: string;
  private _secret: string;

  constructor(options: IDeviceOptions) {
    super(options);
    this._accountId = options.accountId;
    this._name = options.name || null;
    this._pin = {
      signature: options.pin?.signature || null,
      updated: options.pin?.updated || null,
    };
    this._publicKey = options.publicKey;
    this._secret = options.secret || null;
  }

  public get accountId(): string {
    return this._accountId;
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(DeviceEvent.NAME_CHANGED, { name: this._name });
  }

  public get pin(): IDevicePIN {
    return this._pin;
  }
  public set pin(pin: IDevicePIN) {
    this._pin = pin;
    this.addEvent(DeviceEvent.PIN_CHANGED, { pin: this._pin });
  }

  public get publicKey(): string {
    return this._publicKey;
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(DeviceEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== DeviceEvent.CREATED) continue;
      throw new Error("Device has already been created");
    }
    this.addEvent(DeviceEvent.CREATED, {
      accountId: this._accountId,
      name: this._name,
      pin: this._pin,
      publicKey: this._publicKey,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
