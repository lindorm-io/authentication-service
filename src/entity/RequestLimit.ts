import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { GrantType } from "../enum";

export interface IRequestLimit extends IEntity {
  backOffUntil: Date;
  failedTries: number;
  grantType: GrantType;
  subject: string;
}

export interface IRequestLimitOptions extends IEntityBaseOptions {
  backOffUntil?: Date;
  failedTries?: number;
  grantType: GrantType;
  subject: string;
}

export class RequestLimit extends EntityBase implements IRequestLimit {
  private _backOffUntil: Date;
  private _failedTries: number;
  private _grantType: GrantType;
  private _subject: string;

  constructor(options: IRequestLimitOptions) {
    super(options);
    this._backOffUntil = options.backOffUntil || null;
    this._failedTries = options.failedTries || 1;
    this._grantType = options.grantType;
    this._subject = options.subject;
  }

  public get backOffUntil(): Date {
    return this._backOffUntil;
  }
  public set backOffUntil(backOffUntil: Date) {
    this._backOffUntil = backOffUntil;
    this._updated = new Date();
  }

  public get failedTries(): number {
    return this._failedTries;
  }
  public set failedTries(failedTries: number) {
    this._failedTries = failedTries;
    this._updated = new Date();
  }

  public get grantType(): GrantType {
    return this._grantType;
  }

  public get subject(): string {
    return this._subject;
  }

  public create(): void {
    /* intentionally left empty */
  }
}
