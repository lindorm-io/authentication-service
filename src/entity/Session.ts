import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions, TObject } from "@lindorm-io/core";
import { SessionEvent } from "../enum";

export interface ISessionAgent {
  browser: string;
  geoIp: TObject<any>;
  os: string;
  platform: string;
  source: string;
  version: string;
}

export interface ISessionAuthorization {
  codeChallenge: string;
  codeMethod: string;
  deviceChallenge: string;
  email: string;
  id: string;
  otpCode: string;
  redirectUri: string;
  responseType: string;
}

export interface ISession extends IEntity {
  accountId: string;
  agent: ISessionAgent;
  authenticated: boolean;
  authorization: ISessionAuthorization;
  clientId: string;
  deviceId: string;
  expires: Date;
  grantType: string;
  refreshId: string;
  scope: string;
}

export interface ISessionOptions extends IEntityBaseOptions {
  accountId?: string;
  agent?: {
    browser?: string;
    geoIp?: TObject<any>;
    os?: string;
    platform?: string;
    source?: string;
    version?: string;
  };
  authenticated?: boolean;
  authorization: {
    codeChallenge: string;
    codeMethod: string;
    deviceChallenge?: string;
    email: string;
    id: string;
    otpCode?: string;
    redirectUri: string;
    responseType: string;
  };
  clientId: string;
  deviceId?: string;
  expires: Date;
  grantType: string;
  refreshId?: string;
  scope: string;
}

export class Session extends EntityBase implements ISession {
  private _accountId: string;
  private _agent: ISessionAgent;
  private _authenticated: boolean;
  private _authorization: ISessionAuthorization;
  private _clientId: string;
  private _deviceId: string;
  private _expires: Date;
  private _grantType: string;
  private _refreshId: string;
  private _scope: string;

  constructor(options: ISessionOptions) {
    super(options);

    this._accountId = options.accountId || null;
    this._agent = {
      browser: options.agent?.browser || null,
      geoIp: options.agent?.geoIp || null,
      os: options.agent?.os || null,
      platform: options.agent?.platform || null,
      source: options.agent?.source || null,
      version: options.agent?.version || null,
    };
    this._authenticated = options.authenticated || false;
    this._authorization = {
      codeChallenge: options.authorization.codeChallenge,
      codeMethod: options.authorization.codeMethod,
      deviceChallenge: options.authorization.deviceChallenge || null,
      email: options.authorization.email,
      id: options.authorization.id,
      otpCode: options.authorization.otpCode || null,
      redirectUri: options.authorization.redirectUri,
      responseType: options.authorization.responseType,
    };
    this._clientId = options.clientId;
    this._deviceId = options.deviceId || null;
    this._expires = options.expires;
    this._grantType = options.grantType;
    this._refreshId = options.refreshId || null;
    this._scope = options.scope;
  }

  public get accountId(): string {
    return this._accountId;
  }
  public set accountId(accountId: string) {
    this._accountId = accountId;
    this.addEvent(SessionEvent.ACCOUNT_ID_CHANGED, { accountId: this._accountId });
  }

  public get agent(): ISessionAgent {
    return this._agent;
  }

  public get authenticated(): boolean {
    return this._authenticated;
  }
  public set authenticated(authenticated: boolean) {
    this._authenticated = authenticated;
    this.addEvent(SessionEvent.AUTHENTICATED_CHANGED, { authenticated: this._authenticated });
  }

  public get authorization(): ISessionAuthorization {
    return this._authorization;
  }

  public get clientId(): string {
    return this._clientId;
  }

  public get deviceId(): string {
    return this._deviceId;
  }

  public get expires(): Date {
    return this._expires;
  }
  public set expires(expires: Date) {
    this._expires = expires;
    this.addEvent(SessionEvent.EXPIRES_CHANGED, { expires: this._expires });
  }

  public get grantType(): string {
    return this._grantType;
  }

  public get refreshId(): string {
    return this._refreshId;
  }
  public set refreshId(refreshId: string) {
    this._refreshId = refreshId;
    this.addEvent(SessionEvent.REFRESH_ID_CHANGED, { refreshId: this._refreshId });
  }

  public get scope(): string {
    return this._scope;
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== SessionEvent.CREATED) continue;
      throw new EntityCreationError("Session");
    }

    this.addEvent(SessionEvent.CREATED, {
      accountId: this._accountId,
      agent: this._agent,
      authenticated: this._authenticated,
      authorization: this._authorization,
      clientId: this._clientId,
      deviceId: this._deviceId,
      expires: this._expires,
      grantType: this._grantType,
      refreshId: this._refreshId,
      scope: this._scope,
    });
  }
}
