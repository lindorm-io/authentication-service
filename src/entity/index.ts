export * from "./Account";
export * from "./RequestLimit";
export * from "./Session";
