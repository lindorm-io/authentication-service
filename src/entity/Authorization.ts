import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/entity";
import { GrantType } from "../enum";
import { Scope } from "@lindorm-io/jwt";

export interface IAuthorization extends IEntity {
  challengeId: string;
  clientId: string;
  codeChallenge: string;
  codeMethod: string;
  deviceId: string;
  email: string;
  expires: Date;
  grantType: GrantType;
  otpCode: string;
  redirectUri: string;
  responseType: string;
  scope: Array<Scope>;
}

export interface IAuthorizationOptions extends IEntityBaseOptions {
  challengeId?: string;
  clientId: string;
  codeChallenge: string;
  codeMethod: string;
  deviceId?: string;
  email: string;
  expires: Date;
  grantType: GrantType;
  otpCode?: string;
  redirectUri: string;
  responseType: string;
  scope: Array<Scope>;
}

export class Authorization extends EntityBase implements IAuthorization {
  readonly challengeId: string;
  readonly clientId: string;
  readonly codeChallenge: string;
  readonly codeMethod: string;
  readonly deviceId: string;
  readonly email: string;
  readonly expires: Date;
  readonly grantType: GrantType;
  readonly otpCode: string;
  readonly redirectUri: string;
  readonly responseType: string;
  readonly scope: Array<Scope>;

  constructor(options: IAuthorizationOptions) {
    super(options);

    this.challengeId = options.challengeId || null;
    this.clientId = options.clientId;
    this.codeChallenge = options.codeChallenge;
    this.codeMethod = options.codeMethod;
    this.deviceId = options.deviceId || null;
    this.email = options.email;
    this.expires = options.expires;
    this.grantType = options.grantType;
    this.otpCode = options.otpCode || null;
    this.redirectUri = options.redirectUri;
    this.responseType = options.responseType;
    this.scope = options.scope;
  }

  public create(): void {
    /* intentionally left empty */
  }
}
