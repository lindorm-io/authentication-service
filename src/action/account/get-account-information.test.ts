import MockDate from "mockdate";
import { Account } from "../../entity";
import { MOCK_ACCOUNT_OPTIONS, MOCK_UUID } from "../../test/mocks";
import { getAccountInformation } from "./get-account-information";
import { winston } from "../../logger";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));
jest.mock("../../support", () => ({
  getAccount: jest.fn(() => () => new Account(MOCK_ACCOUNT_OPTIONS)),
  getAccountDevices: jest.fn(() => () => ["devices"]),
  getAccountSessions: jest.fn(() => () => ["sessions"]),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("getAccountInformation", () => {
  let getMockContext: any;

  beforeEach(() => {
    getMockContext = () => ({
      logger: winston,
    });
  });

  test("should return account information", async () => {
    await expect(
      getAccountInformation(getMockContext())({
        accountId: MOCK_UUID,
      }),
    ).resolves.toMatchSnapshot();
  });
});
