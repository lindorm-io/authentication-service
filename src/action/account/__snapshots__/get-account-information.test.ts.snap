// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAccountInformation should return account information 1`] = `
Object {
  "created": 2020-01-01T07:00:00.000Z,
  "devices": Array [
    "devices",
  ],
  "email": "email@lindorm.io",
  "hasOtp": false,
  "hasPassword": false,
  "permission": "user",
  "sessions": Array [
    "sessions",
  ],
  "updated": 2020-01-01T07:00:00.000Z,
}
`;
